lib/codegen/ProofEvent.cpp

```diff
@@ -0,0 +1,198 @@
+          llvm::Type::getVoidTy(Module->getContext()),
+          llvm::Type::getInt8PtrTy(Module->getContext()),
+          llvm::Type::getInt8PtrTy(Module->getContext())),
+      {outputFile, nameptr});
+
+  llvm::BranchInst::Create(MergeBlock, TrueBlock);
+
+  // Some places will try to use llvm::Instruction::insertAfter on the back of
+  // the MergeBlock.
+  // If the MergeBlock has no instructions, this has resulted in a segfault when
+  // printing the IR. Adding an effective nop prevents this.
+  llvm::BinaryOperator::Create(
+      llvm::Instruction::Add,
+      llvm::ConstantInt::get(llvm::Type::getInt8Ty(Module->getContext()), 0),
+      llvm::ConstantInt::get(llvm::Type::getInt8Ty(Module->getContext()), 0),
+      "nop", MergeBlock);
+
+  return CurrentBlock = MergeBlock;
+}
+
+llvm::BasicBlock *
+ProofEvent::hookEvent_post(llvm::Value *val, KORECompositeSort *sort) {
+  auto b = proofBranch("hookpost");
+  auto TrueBlock = b.first;
+  auto MergeBlock = b.second;
+  auto OutputFileName = Module->getOrInsertGlobal(
+      "output_file", llvm::Type::getInt8PtrTy(Module->getContext()));
+  auto ir = new llvm::IRBuilder(TrueBlock);
+  auto outputFile = new llvm::LoadInst(
+      llvm::Type::getInt8PtrTy(Module->getContext()), OutputFileName, "output",
+      TrueBlock);
+
+  auto cat = sort->getCategory(Definition);
+  std::ostringstream Out;
+  sort->print(Out);
+  auto sortptr = ir->CreateGlobalStringPtr(Out.str(), "", 0, Module);
+
+  writeUInt64(outputFile, Module, 0xbbbbbbbbbbbbbbbb, TrueBlock);
+  if (cat.cat == SortCategory::Symbol || cat.cat == SortCategory::Variable) {
+    ir->CreateCall(
+        getOrInsertFunction(
+            Module, "serializeTermToFile",
+            llvm::Type::getVoidTy(Module->getContext()),
+            llvm::Type::getInt8PtrTy(Module->getContext()),
+            getValueType({SortCategory::Symbol, 0}, Module),
+            llvm::Type::getInt8PtrTy(Module->getContext())),
+        {outputFile, val, sortptr});
+  } else if (val->getType()->isIntegerTy()) {
+    val = ir->CreateIntToPtr(
+        val, llvm::Type::getInt8PtrTy(Module->getContext()));
+    ir->CreateCall(
+        getOrInsertFunction(
+            Module, "serializeRawTermToFile",
+            llvm::Type::getVoidTy(Module->getContext()),
+            llvm::Type::getInt8PtrTy(Module->getContext()),
+            llvm::Type::getInt8PtrTy(Module->getContext()),
+            llvm::Type::getInt8PtrTy(Module->getContext())),
+        {outputFile, val, sortptr});
+  } else {
+    val = ir->CreatePointerCast(
+        val, llvm::Type::getInt8PtrTy(Module->getContext()));
+    ir->CreateCall(
+        getOrInsertFunction(
+            Module, "serializeRawTermToFile",
+            llvm::Type::getVoidTy(Module->getContext()),
+            llvm::Type::getInt8PtrTy(Module->getContext()),
+            llvm::Type::getInt8PtrTy(Module->getContext()),
+            llvm::Type::getInt8PtrTy(Module->getContext())),
+        {outputFile, val, sortptr});
+  }
```

Comment:
This particular idiom to safely serialize a term given its sort and type gets repeated below; could we pull it out into its own function? The changes in #862 will need the same code as well, so it makes sense to extract it.

Reply:
Yep, this block here appears in multiple places and should definitely be factored into its own function.